persist/fs/commitlog/reader.go

```diff
@@ -243,19 +252,19 @@ func (r *reader) decoderLoop(inBuf <-chan decoderArg, outBuf chan<- *readRespons
 			namespace.IncRef()
 			namespace.AppendAll(decoded.Namespace)
 
-			r.metadataLock.Lock()
-			r.metadataLookup[entry.Index] = Series{
+			r.metadata.Lock()
+			r.metadata.lookup[entry.Index] = Series{
 				UniqueIndex: entry.Index,
 				ID:          ts.BinaryID(id),
 				Namespace:   ts.BinaryID(namespace),
 				Shard:       decoded.Shard,
 			}
-			waitGroup, ok := r.metadataWaitGroups[entry.Index]
+			waitGroup, ok := r.metadata.wgs[entry.Index]
 			// Another goroutine is blocked waiting for this metadata
 			if ok {
 				waitGroup.Done()
```

Comment:
You probably want to wait outside holding the lock.  In fact this seems like a deadlock scenario, do we have a test for this branch?

How can you wait for the metadata to be available in the map while you're holding the lock if one can only put it in the map by acquiring the lock?

Reply:
I'm a little confused here,  waitgroup.Done() notifies all the goroutines that are waiting for this metadata that its now available, this part of the code is not waiting for the metadata to be available in the map, its the opposite its puting the metadata in the map and notifying everyone else that the data is available.

the call to waitGroup.Wait() is called on 301 and that method is invoked while the worker is not holding any locks.